src/client/libds3/bucket.c

```diff
@@ -7,31 +7,343 @@
+	if (rc != 0) {
+		D_ERROR("Failed to open to container %s, rc = %d\n", name, rc);
+		rc = daos_der2errno(rc);
+		goto err_ds3b;
+	}
+
+	rc = dfs_mount(ds3->poh, ds3b_tmp->coh, O_RDWR, &ds3b_tmp->dfs);
+
+	if (rc != 0) {
+		goto err;
+	}
+
+	*ds3b = ds3b_tmp;
 	return 0;
+
+err:
+	daos_cont_close(ds3b_tmp->coh, ev);
+err_ds3b:
+	D_FREE(ds3b);
+	return -rc;
 }
 
 int
 ds3_bucket_close(ds3_bucket_t *ds3b, daos_event_t *ev)
 {
-	return 0;
+	int rc = 0;
+	rc     = dfs_umount(ds3b->dfs);
+	daos_cont_close(ds3b->coh, ev);
+	D_FREE(ds3b);
+	return -rc;
+}
+
+int
+ds3_bucket_get_info(struct ds3_bucket_info *info, ds3_bucket_t *ds3b, daos_event_t *ev)
+{
+	if (ds3b == NULL || info == NULL)
+		return -EINVAL;
+
+	char const *const names[]  = {RGW_BUCKET_INFO};
+	void *const       values[] = {info->encoded};
+	size_t            sizes[]  = {info->encoded_length};
+	int               rc       = daos_cont_get_attr(ds3b->coh, 1, names, values, sizes, ev);
+	rc                         = -daos_der2errno(rc);
+	return rc;
+}
+
+int
+ds3_bucket_set_info(struct ds3_bucket_info *info, ds3_bucket_t *ds3b, daos_event_t *ev)
+{
+	if (ds3b == NULL || info == NULL)
+		return -EINVAL;
+
+	char const *const names[]  = {RGW_BUCKET_INFO};
+	void const *const values[] = {info->encoded};
+	size_t const      sizes[]  = {info->encoded_length};
+	int               rc       = daos_cont_set_attr(ds3b->coh, 1, names, values, sizes, ev);
+	rc                         = -daos_der2errno(rc);
+	return rc;
+}
+
+int
+ds3_bucket_list_obj(uint32_t *nobj, struct ds3_object_info *objs, uint32_t *ncp,
+		    struct ds3_common_prefix_info *cps, const char *prefix, const char *delim,
+		    char *marker, bool list_versions, bool *is_truncated, ds3_bucket_t *ds3b)
+{
+	if (ds3b == NULL || nobj == NULL)
+		return -EINVAL;
+
+	// End
+	if (*nobj == 0)
+		return 0;
+
+	// TODO: support the case when delim is not /
+	if (strcmp(delim, "/") != 0) {
+		return -EINVAL;
+	}
+
+	int         rc          = 0;
+	char       *file_start  = strrchr(prefix, delim[0]);
+	const char *path        = "";
+	const char *prefix_rest = prefix;
+
+	if (file_start != NULL) {
+		*file_start = '\0';
+		path        = prefix;
+		prefix_rest = file_start + 1;
+	}
+
+	dfs_obj_t *dir_obj;
+
+	char      *lookup_path;
+	D_ALLOC_ARRAY(lookup_path, DS3_MAX_KEY);
+	if (lookup_path == NULL)
+		return -ENOMEM;
+
+	strcpy(lookup_path, "/");
+	strcat(lookup_path, path);
+	rc = dfs_lookup(ds3b->dfs, lookup_path, O_RDWR, &dir_obj, NULL, NULL);
+
+	if (rc != 0) {
+		goto err_path;
+	}
+
+	struct dirent *dirents;
+	D_ALLOC_ARRAY(dirents, *nobj);
+	if (dirents == NULL) {
+		rc = ENOMEM;
+		goto err_dir_obj;
+	}
+
+	// TODO handle bigger directories
+	// TODO handle ordering
+	// TODO handle marker
+	daos_anchor_t anchor;
+	daos_anchor_init(&anchor, 0);
+
+	rc = dfs_readdir(ds3b->dfs, dir_obj, &anchor, nobj, dirents);
+	if (rc != 0) {
+		goto err_dirents;
+	}
+
+	if (is_truncated != NULL) {
+		*is_truncated = !daos_anchor_is_eof(&anchor);
```

Comment:
you should keep iterating until you hit nobj entries, no?
you could request 20 entries for example, but get 10 on first call (eof has not been hit yet), so you will need to call readdir again.

Reply:
The implementation of `dfs_readdir` does that for us. But it might be better to also include it, added a task